framework/aster-frame/src/task/scheduler.rs

```diff
@@ -1,62 +1,146 @@
 use crate::prelude::*;
-use crate::sync::SpinLock;
-use crate::task::Task;
+use crate::sync::{SpinLock, SpinLockGuard};
+use crate::task::{NeedResched, ReadPriority, Task};
 
-use lazy_static::lazy_static;
+use log::debug;
 
-lazy_static! {
-    pub(crate) static ref GLOBAL_SCHEDULER: SpinLock<GlobalScheduler> =
-        SpinLock::new(GlobalScheduler { scheduler: None });
+pub(crate) static GLOBAL_SCHEDULER: spin::Once<SpinLock<GlobalScheduler>> = spin::Once::new();
+
+pub fn init() {
+    GLOBAL_SCHEDULER.call_once(|| SpinLock::new(GlobalScheduler { scheduler: None }));
 }
 
 /// A scheduler for tasks.
 ///
-/// An implementation of scheduler can attach scheduler-related information
-/// with the `TypeMap` returned from `task.data()`.
-pub trait Scheduler: Sync + Send {
-    fn enqueue(&self, task: Arc<Task>);
+/// Operations on the scheduler should be performed with interrupts disabled,
+/// which has been ensured by the callers of the `GLOBAL_SCHEDULER`.
+/// Therefore, implementations of this trait do not need to worry about interrupt safety.
+pub trait Scheduler<T: NeedResched + ReadPriority = Task>: Sync + Send {
+    /// Called when a task enters a runnable state.
+    /// Add the task to the scheduler.
+    fn enqueue(&self, task: Arc<T>);
+
+    /// Called when the task is no longer alive.
+    /// Remove the task-related from the scheduler.
+    ///
+    /// A similar method in the linux scheduler is `dequeue_task`.
+    ///
+    /// Return `true` if the task was in the scheduler.
+    fn remove(&self, task: &Arc<T>) -> bool;
+
+    /// Choose the most appropriate task eligible to run next.
+    fn pick_next_task(&self) -> Option<Arc<T>>;
+
+    /// Tells whether the given current task should be preempted
+    /// by other tasks in the queue.
+    fn should_preempt(&self, cur_task: &Arc<T>) -> bool;
+
+    /// Handle a tick from the timer.
+    /// Modify the states of the given task(the current task in `PROCESSOR`)
+    /// according to the time update.
+    ///
+    /// # Arguments
+    ///
+    /// * `cur_task` - the task to be charged, must be held by the processor, and not in runqueue
+    ///
+    /// # Returns
+    ///
+    /// `true` if the `task` need a rescudule.
+    fn tick(&self, cur_task: &Arc<T>) -> bool;
```

Comment:
I don't get why we need the return value and where we are using it currently.

Reply:
Yes, it's of no use anymore and will be discarded.